daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/ExceptionTest.scala

```diff
@@ -0,0 +1,536 @@
+                        catch e ->
+                            throw @(Option (Update Int64)) @M:MyException (M:MyException {message = "oops"}) //CHANGE
+                    in
+                        upure @Int64 (ADD_INT64 200 x)
+                catch e ->
+                    Some @(Update Int64) (upure @Int64 88)
+            in
+                upure @Int64 (ADD_INT64 100 x)
+         ;
+
+       }
+      """)
+
+    val testCases = Table[String, Long](
+      ("expression", "expected"),
+      ("M:innerCatch", 377),
+      ("M:outerCatch", 188),
+      ("M:throwWhileInnerHandlerDecides", 188),
+    )
+
+    forEvery(testCases) { (exp: String, num: Long) =>
+      s"eval[$exp] --> $num" in {
+        val expected: SResult = SResultFinalValue(SValue.SInt64(num))
+        runUpdateExpr(pkgs)(e"$exp") shouldBe expected
+      }
+    }
+  }
+
+  "selective exception handling (using payload)" should {
+
+    val pkgs: PureCompiledPackages = typeAndCompile(p"""
+      module M {
+        record @serializable MyException = { payload: Int64 } ;
+
+        exception MyException = {
+          message \(e: M:MyException) -> "no-message"
+        };
+
+        val maybeInnerCatch : Bool -> Bool -> Update Int64 = \(catchAtInner: Bool) (throwAtInner: Bool) ->
+          ubind x: Int64 <-
+            try @Int64
+              ubind x: Int64 <-
+                try @Int64
+                  upure @Int64 (throw @Int64 @M:MyException (M:MyException {payload = 1000}))
+                catch e ->
+                  case catchAtInner of
+                    True ->
+                      (case (from_any_exception @M:MyException e) of
+                        None -> None @(Update Int64)
+                      | Some my -> Some @(Update Int64) (upure @Int64 (ADD_INT64 77 (M:MyException {payload} my))))
+                  | False ->
+                      (case throwAtInner of
+                         True -> throw @(Option (Update Int64)) @M:MyException (M:MyException {payload = 2000})
+                       | False -> None @(Update Int64))
+              in
+                upure @Int64 (ADD_INT64 200 x)
+            catch e ->
+              case (from_any_exception @M:MyException e) of
+                None -> None @(Update Int64)
+              | Some my -> Some @(Update Int64) (upure @Int64 (ADD_INT64 88 (M:MyException {payload} my)))
+
+          in
+            upure @Int64 (ADD_INT64 100 x) ;
+       }
+      """)
+
+    val testCases = Table[String, Long](
+      ("expression", "expected"),
+      ("M:maybeInnerCatch True False", 1377),
+      ("M:maybeInnerCatch False False", 1188),
+      ("M:maybeInnerCatch False True", 2188),
+    )
+
+    forEvery(testCases) { (exp: String, num: Long) =>
+      s"eval[$exp] --> $num" in {
+        val expected: SResult = SResultFinalValue(SValue.SInt64(num))
+        runUpdateExpr(pkgs)(e"$exp") shouldBe expected
+      }
+    }
+  }
+
+  "throw/catch (control flow)" should {
```

Comment:
some comments on what the tests are supposed to check for would be super useful here. Decyphering hand-written daml-lf always is a bit tricky.

Reply:
They were hard to write... they should be hard to read :)

Reply:
I added a bunch of comments which hopefully reveal the intent of the various tests.